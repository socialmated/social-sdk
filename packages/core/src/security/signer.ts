import { type Options } from 'got';

export interface Signer<T> {
  sign: ((req: Options) => Promise<T>) & ((req: Request) => Promise<T>);
}
