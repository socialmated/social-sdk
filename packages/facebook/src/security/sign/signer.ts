import { type Signer } from '@social-sdk/client/security';
import { type Options } from '@social-sdk/client/http';
import { generateProof, type AppSecretProof } from './sign.js';

export class AppSecretProofSigner implements Signer<AppSecretProof> {
  constructor(private readonly appSecret: string) {}

  public sign(req: Options): AppSecretProof {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- should be defined in the request
    const accessToken = new URLSearchParams(new URL(req.url!).searchParams).get('access_token');
    if (!accessToken) {
      throw new Error('Access token is required to generate App Secret Proof');
    }

    return generateProof(this.appSecret, accessToken);
  }
}
