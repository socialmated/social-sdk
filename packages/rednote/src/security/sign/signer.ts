import { type CookieSession } from '@social-sdk/core/auth/session';
import { type Options } from 'got';
import { type Signer } from '@social-sdk/core/security';
import { generateLocalId } from '../token/token.js';
import { signOld, signNew, generateMns, type XhsSignOutput } from './sign.js';
import { defaultConfig } from '@/client/config.js';

/**
 * The output type for the XHS signing process.
 */
interface XhsSignerOptions {
  /**
   * The type of signing to perform.
   */
  type?: 'sign_new' | 'sign_old';

  /**
   * Whether MNS is supported.
   */
  mns?: boolean;
}

/**
 * The `XhsSigner` class provides methods to sign requests for the Xiaohongshu (XHS) API.
 *
 * @example
 * ```typescript
 * const signer = XhsSigner.fromSession(session);
 * const signature = await signer.sign('/api/resource', requestBody, { type: 'sign_new' });
 * ```
 */
class XhsSigner implements Signer<{ 'X-s': string; 'X-t': string; 'X-Mns': string }> {
  /**
   * The `a1` cookie value.
   */
  private a1: string;

  /**
   * Creates an instance of `XhsSigner`.
   * @param a1 - The `a1` cookie value used for signing requests.
   */
  constructor(a1?: string) {
    this.a1 = a1 ?? generateLocalId(defaultConfig.platform);
  }

  /**
   * Creates an instance of `XhsSigner` from a given session.
   *
   * @param session - The `CookieSession` object containing authentication cookies.
   * @returns An instance of `XhsSigner` initialized with the `a1` cookie value from the session.
   * @throws If the session does not contain an `a1` cookie.
   */
  public static fromSession(session: CookieSession): XhsSigner {
    const a1 = session.get('a1');
    if (!a1) {
      throw new Error('Session does not contain a1 cookie');
    }

    return new XhsSigner(a1);
  }

  /**
   * Generates a signature for the given request.
   *
   * @param request - The request options.
   * @param options - Optional parameters for signing, including type and MNS support.
   * @returns An object containing the signed headers, including 'X-s', 'X-t', and 'X-Mns'.
   */
  public sign(request: Options, options?: XhsSignerOptions): XhsSignOutput {
    const fullPath = request.url ? new URL(request.url).pathname + new URL(request.url).search : '';
    const body: unknown =
      !request.json && typeof request.body === 'string' ? JSON.parse(request.body) : (request.json ?? request.body);
    let output: Omit<XhsSignOutput, 'X-Mns'>;

    if (options?.type === 'sign_old') {
      output = signOld(fullPath, body);
    } else {
      output = signNew(fullPath, body, this.a1);
    }

    return {
      ...output,
      'X-Mns': options?.mns ? generateMns(fullPath, body) : 'unload',
    };
  }
}

export { XhsSigner };
export type { XhsSignerOptions };
