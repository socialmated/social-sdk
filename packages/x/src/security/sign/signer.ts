import { type Options } from 'got';
import { type Signer } from '@social-sdk/core/security';
import { HomeHtmlParser, OnDemandJsParser } from './parser.js';
import { transactionId, deriveAnimationKey, decodeVerificationKey } from './sign.js';

/**
 * Generates and signs the transaction ID for X's API requests.
 *
 * @example
 * ```typescript
 * const signer = new TransactionIdSigner();
 * const transactionId = await signer.sign('POST', '/i/api/1.1/some-endpoint.json');
 * console.log(transactionId);
 * ```
 *
 * @public
 */
class TransactionIdSigner implements Signer<string> {
  /**
   * The cryptographic verification key extracted from the homepage.
   */
  private key?: Uint8Array;

  /**
   * The animation key derived from homepage animation data.
   */
  private animationKey?: string;

  /**
   * Generates and signs a transaction ID for a given HTTP method and API endpoint path.
   *
   * @param method - The HTTP method (e.g., 'GET', 'POST').
   * @param path - The API endpoint path (e.g., '/api/endpoint').
   * @param options - The got Options object.
   * @returns Promise resolving to a Base64-encoded transaction ID string.
   * @throws Error if the signer has not been initialized.
   */
  public async sign(method: string, path: string): Promise<string>;
  /**
   * Generates and signs a transaction ID for a `got` request {@link Options}.
   *
   * @param opts - The `got` request {@link Options} object.
   * @returns Promise resolving to a Base64-encoded transaction ID string.
   * @throws Error if the signer has not been initialized.
   */
  public async sign(opts: Options): Promise<string>;
  /**
   * Generates and signs a transaction ID for a `fetch` {@link Request}.
   *
   * @param req - The `fetch` request {@link Request} object.
   * @returns Promise resolving to a Base64-encoded transaction ID string.
   * @throws Error if the signer has not been initialized or if invalid arguments are provided.
   */
  // eslint-disable-next-line @typescript-eslint/unified-signatures -- explicitly overload
  public async sign(req: Request): Promise<string>;
  public async sign(arg1: string | Options | Request, arg2?: string): Promise<string> {
    if (!this.key || !this.animationKey) {
      await this.init();
    }

    let method: string;
    let requestPath: string;
    if (typeof arg1 === 'string') {
      method = arg1;
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- expected to be defined at this point
      requestPath = arg2!;
    } else {
      method = arg1.method;
      requestPath = arg1.url ? new URL(arg1.url).pathname : '';
    }

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- expected to be defined at this point
    return transactionId(method, requestPath, this.key!, this.animationKey!);
  }

  /**
   * Initializes the instance by extracting cryptographic keys and animation data.
   * This method must be called before signing transaction IDs.
   *
   * @returns Promise that resolves when initialization is complete.
   * @throws Error if required data cannot be extracted.
   */
  private async init(): Promise<void> {
    if (this.key !== undefined && this.animationKey !== undefined) {
      // Already initialized
      return;
    }

    // Create parsers for the homepage and ondemand JavaScript file.
    const htmlParser = await HomeHtmlParser.create();
    const jsParser = await OnDemandJsParser.create(htmlParser.findOnDemandJsChunkHash());

    // Extract the verification key from the homepage.
    const verificationKey = htmlParser.getVerificationKey();
    if (!verificationKey) {
      throw new Error("Couldn't extract verification key from the homepage");
    }
    this.key = decodeVerificationKey(verificationKey);

    // Derive the animation key from SVG frames and specific indices.
    const frames = htmlParser.getAnimationFrames();
    const indices = jsParser.getIndices();
    this.animationKey = deriveAnimationKey(this.key, indices, frames);
  }
}

export { TransactionIdSigner };
