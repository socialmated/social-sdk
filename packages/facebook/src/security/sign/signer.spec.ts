import { Options } from 'got';
import { AppSecretProofSigner } from './signer.js';

describe(AppSecretProofSigner, () => {
  const appSecret = 'test_app_secret_12345';
  const accessToken = 'test_access_token_abcdef';

  let signer: AppSecretProofSigner;

  beforeEach(() => {
    vi.setSystemTime(new Date(2025, 0, 1));
    signer = new AppSecretProofSigner(appSecret);
  });

  describe(AppSecretProofSigner.prototype.sign, () => {
    it('should sign request with access token in URL and return app secret proof', () => {
      const req = new Options({
        method: 'GET',
        url: `https://graph.facebook.com/v12.0/me?access_token=${accessToken}`,
      });
      const proof = signer.sign(req);

      expect(proof).toMatchSnapshot();
    });

    it('should sign request with access token in searchParams string and return app secret proof', () => {
      const req = new Options({
        method: 'GET',
        url: 'https://graph.facebook.com/v12.0/me',
        searchParams: `access_token=${accessToken}&fields=id,name`,
      });
      const proof = signer.sign(req);

      expect(proof).toMatchSnapshot();
    });

    it('should sign request with access token in searchParams URLSearchParams and return app secret proof', () => {
      const searchParams = new URLSearchParams();
      searchParams.set('access_token', accessToken);
      searchParams.set('fields', 'id,name');

      const req = new Options({
        method: 'GET',
        url: 'https://graph.facebook.com/v12.0/me',
        searchParams,
      });
      const proof = signer.sign(req);

      expect(proof).toMatchSnapshot();
    });

    it('should sign request with access token in searchParams object and return app secret proof', () => {
      const req = new Options({
        method: 'POST',
        url: 'https://graph.facebook.com/v12.0/me/photos',
        searchParams: {
          access_token: accessToken,
          message: 'Test photo upload',
        },
      });
      const proof = signer.sign(req);

      expect(proof).toMatchSnapshot();
    });

    it('should prioritize access token from URL over searchParams when both are present', () => {
      const urlAccessToken = 'url_access_token_123';
      const req = new Options({
        method: 'GET',
        url: `https://graph.facebook.com/v12.0/me?access_token=${urlAccessToken}`,
        searchParams: {
          access_token: accessToken,
        },
      });
      const proof = signer.sign(req);

      expect(proof).toMatchSnapshot();
    });

    it('should throw error when access token is not provided', () => {
      const req = new Options({
        method: 'GET',
        url: 'https://graph.facebook.com/v12.0/me',
      });

      expect(() => signer.sign(req)).toThrow('Access token is required to generate App Secret Proof');
    });

    it('should throw error when access token is null in searchParams object', () => {
      const req = new Options({
        method: 'GET',
        url: 'https://graph.facebook.com/v12.0/me',
        searchParams: {
          access_token: null,
          fields: 'id,name',
        },
      });

      expect(() => signer.sign(req)).toThrow('Access token is required to generate App Secret Proof');
    });
  });
});
