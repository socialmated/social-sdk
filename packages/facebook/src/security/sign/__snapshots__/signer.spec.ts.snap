// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppSecretProofSigner > sign > should prioritize access token from URL over searchParams when both are present 1`] = `
{
  "proof": "29346adb4c7a5f3e9116208a19d2804658cd9bd3ac2f2783a7622f3970e540a0",
  "time": 1735657200000,
}
`;

exports[`AppSecretProofSigner > sign > should sign request with access token in URL and return app secret proof 1`] = `
{
  "proof": "29346adb4c7a5f3e9116208a19d2804658cd9bd3ac2f2783a7622f3970e540a0",
  "time": 1735657200000,
}
`;

exports[`AppSecretProofSigner > sign > should sign request with access token in searchParams URLSearchParams and return app secret proof 1`] = `
{
  "proof": "29346adb4c7a5f3e9116208a19d2804658cd9bd3ac2f2783a7622f3970e540a0",
  "time": 1735657200000,
}
`;

exports[`AppSecretProofSigner > sign > should sign request with access token in searchParams object and return app secret proof 1`] = `
{
  "proof": "29346adb4c7a5f3e9116208a19d2804658cd9bd3ac2f2783a7622f3970e540a0",
  "time": 1735657200000,
}
`;

exports[`AppSecretProofSigner > sign > should sign request with access token in searchParams string and return app secret proof 1`] = `
{
  "proof": "29346adb4c7a5f3e9116208a19d2804658cd9bd3ac2f2783a7622f3970e540a0",
  "time": 1735657200000,
}
`;
